content/blog/2022/2022-01-25-annotation-processing.md

```diff
@@ -0,0 +1,951 @@
+
+```java
+@Company
+public class Employee {
+
+    @Company
+    public void getEmployeeStatus(){
+
+        System.out.println("This is the Base Employee class");
+    }
+}
+```
+
+If we apply @Company to the method 'getEmployeeStatus' as above, we get a compiler error stating: "'@Company' not applicable to method."
+
+The various self-explanatory Target types are:
+
+​		ElementType.ANNOTATION_TYPE, 
+​		ElementType.CONSTRUCTOR
+​		ElementType.FIELD
+​		ElementType.LOCAL_VARIABLE
+​		ElementType.METHOD
+​		ElementType.PACKAGE
+​		ElementType.PARAMETER
+​		ElementType.TYPE 
+
+#### @Retention
+
+@Retention specifies when the annotation is discarded.
+
+​	  SOURCE - The annotation is used at compile-time and discarded at run-time.
+
+​	  CLASS - The annotation is stored in the class file at compile time and discarded at run time.
+
+​	  RUNTIME - The annotation is retained at run-time.
+
+If we needed an annotation to only provide error checking at compile-time like @Override we would use SOURCE. If we need an annotation to provide functionality at run-time such as @Test in Junit we would use RUNTIME. To see a real example, create the following annotations in 3 separate files:
+
+```java
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.CLASS)
+public @interface ClassRetention {
+}
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.SOURCE)
+public @interface SourceRetention {
+}
+
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface RuntimeRetention {
+}
+```
+
+Now create a class that uses all 3 annotations:
+
+```java
+@SourceRetention
+@RuntimeRetention
+@ClassRetention
+public class EmployeeRetentionAnnotation {
+}
+```
+
+To verify that only the run-time annotation is available at run-time, run a test as follows:
+
+```java
+public class RetentionTest {
+
+    public static void main(String[] args) {
+
+        SourceRetention[] sourceRetention = new EmployeeRetentionAnnotation().getClass().getAnnotationsByType(SourceRetention.class);
+        System.out.println("Source Retentions at run-time: " + sourceRetention.length);
+
+        RuntimeRetention[] runtimeRetention = new EmployeeRetentionAnnotation().getClass().getAnnotationsByType(RuntimeRetention.class);
+        System.out.println("Run-time Retentions at run-time: " + runtimeRetention.length);
+
+        ClassRetention[] classRetention = new EmployeeRetentionAnnotation().getClass().getAnnotationsByType(ClassRetention.class);
+        System.out.println("Class Retentions at run-time: " + classRetention.length);
+    }
+}
+```
+
+The output would be as follows:
+
+```
+Source Retentions at run-time: 0
+Run-time Retentions at run-time: 1
+Class Retentions at run-time: 0
+```
+
+So we verified that only the RUNTIME annotation gets processed at run-time.
+
+## Annotation Categories
```

Comment:
This section is missing an introduction. Why does it make sense to categorize annotations? Why should the reader read this section?